flowmachine/flowmachine/features/subscriber/distance_series.py

```diff
@@ -0,0 +1,140 @@
+Per subscriber time series of distances from some reference location.
+"""
+from typing import List, Optional
+
+from flowmachine.features.spatial import DistanceMatrix
+from .metaclasses import SubscriberFeature
+from ..utilities.subscriber_locations import SubscriberLocations, BaseLocation
+
+
+valid_stats = {"sum", "avg", "max", "min", "median", "stddev", "variance"}
+valid_time_buckets = ["second", "minute", "hour", "day", "month", "year", "century"]
+
+
+class DistanceSeries(SubscriberFeature):
+    """
+    Per subscriber time series of distance in meters from some reference location.
+
+    Parameters
+    ----------
+    subscriber_locations : SubscriberLocations
+        A subscriber locations query with a lon-lat spatial unit to build the distance series against.
+    reference_location : BaseLocation or None, default None
+        The set of home locations from which to calculate distance at each sighting.
+        If not given then distances will be from (0, 0).
+    statistic : str
+        the statistic to calculate one of 'sum', 'avg', 'max', 'min', 
+        'median', 'stddev' or 'variance'
+    time_bucket : {"second", "minute", "hour", "day", "month", "year", "century"}, default "day"
+        Time bucket to calculate the statistic over.
+
+    Examples
+    --------
+    >>> d = DistanceSeries(subscriber_locations=SubscriberLocations("2016-01-01", "2016-01-07", spatial_unit=make_spatial_unit("lon-lat")))
+    >>> d.head()
+             subscriber    datetime         value
+    0  038OVABN11Ak4W5P  2016-01-01  9.384215e+06
+    1  038OVABN11Ak4W5P  2016-01-02  9.233302e+06
+    2  038OVABN11Ak4W5P  2016-01-03  9.376996e+06
+    3  038OVABN11Ak4W5P  2016-01-04  9.401404e+06
+    4  038OVABN11Ak4W5P  2016-01-05  9.357210e+06
+    """
+
+    def __init__(
+        self,
+        *,
+        subscriber_locations: SubscriberLocations,
+        reference_location: Optional[BaseLocation] = None,
+        statistic: str = "avg",
+        time_bucket: str = "day",
+    ):
+        subscriber_locations.spatial_unit.verify_criterion("has_geography")
+        subscriber_locations.spatial_unit.verify_criterion("has_lon_lat_columns")
+        self.spatial_unit = subscriber_locations.spatial_unit
+        if time_bucket.lower() in valid_time_buckets:
+            self.aggregate_by = time_bucket.lower()
+        else:
+            raise ValueError(
+                f"'{time_bucket}' is not a valid value for time_bucket. Use one of {valid_time_buckets}"
+            )
+
+        self.statistic = statistic.lower()
+        if self.statistic not in valid_stats:
+            raise ValueError(
+                "{} is not a valid statistic. Use one of {}".format(
+                    self.statistic, valid_stats
+                )
+            )
+        self.start = subscriber_locations.start
+        self.stop = subscriber_locations.stop
+        if reference_location is None:
+            # Using 0, 0
+            self.reference_location = None
```

Comment:
Is there a reason why we can't directly say something like `self.reference_location = (0, 0)` here? (Not sure what the right syntax is to create a point location, but you get the idea.)

Or does this prevent us from doing some sort of optimisation later because the value is constant?

Reply:
(I would hope that setting the `reference_location` here would allow us to get rid of the if/else branch in `_make_query()` below.)